plonky2x/core/src/frontend/merkle/tendermint.rs

```diff
@@ -8,13 +9,58 @@ use crate::frontend::merkle::utils::log2_ceil_usize;
 use crate::frontend::vars::Bytes32Variable;
 use crate::prelude::{
     ArrayVariable, BoolVariable, ByteVariable, BytesVariable, CircuitBuilder, CircuitVariable,
+    Variable,
 };
 
+pub trait TendermintMerkleTree {
+    fn get_root_from_merkle_proof_hashed_leaf<const PROOF_DEPTH: usize>(
+        &mut self,
+        proof: &ArrayVariable<Bytes32Variable, PROOF_DEPTH>,
+        path_indices: &ArrayVariable<BoolVariable, PROOF_DEPTH>,
+        leaf: Bytes32Variable,
+    ) -> Bytes32Variable;
+
+    fn get_root_from_merkle_proof<const PROOF_DEPTH: usize, const LEAF_SIZE_BYTES: usize>(
+        &mut self,
+        inclusion_proof: &MerkleInclusionProofVariable<PROOF_DEPTH, LEAF_SIZE_BYTES>,
+        path_indices: &ArrayVariable<BoolVariable, PROOF_DEPTH>,
+    ) -> Bytes32Variable;
+
+    fn leaf_hash(&mut self, leaf: &[ByteVariable]) -> Bytes32Variable;
+
+    fn inner_hash(&mut self, left: &Bytes32Variable, right: &Bytes32Variable) -> Bytes32Variable;
+
+    fn hash_merkle_layer(
+        &mut self,
+        merkle_hashes: Vec<Bytes32Variable>,
+        merkle_hash_enabled: Vec<BoolVariable>,
+    ) -> (Vec<Bytes32Variable>, Vec<BoolVariable>);
+
+    fn hash_leaves<const LEAF_SIZE_BYTES: usize>(
+        &mut self,
+        leaves: Vec<BytesVariable<LEAF_SIZE_BYTES>>,
+    ) -> Vec<Bytes32Variable>;
+
+    fn get_root_from_hashed_leaves<const MAX_NB_LEAVES: usize>(
+        &mut self,
+        leaf_hashes: ArrayVariable<Bytes32Variable, MAX_NB_LEAVES>,
+        nb_enabled_leaves: Variable,
+    ) -> Bytes32Variable;
+
+    fn compute_root_from_leaves<const MAX_NB_LEAVES: usize, const LEAF_SIZE_BYTES: usize>(
+        &mut self,
+        leaves: ArrayVariable<BytesVariable<LEAF_SIZE_BYTES>, MAX_NB_LEAVES>,
+        nb_enabled_leaves: Variable,
+    ) -> Bytes32Variable;
+}
+
 /// Merkle Tree implementation for the Tendermint spec (follows Comet BFT Simple Merkle Tree spec: https://docs.cometbft.com/main/spec/core/encoding#merkle-trees).
+/// Adds pre-image prefix of 0x01 to inner nodes and 0x00 to leaf nodes for second pre-image resistance.
+/// Computed root hash is independent of the number of empty leaves, unlike the simple Merkle Tree.
 /// TODO: Create generic interface for Merkle trees to implement.
```

Comment:
nit: should we remove this comment? it seems like you went with 2 separate traits?